oauth/api/router.py

```diff
@@ -0,0 +1,167 @@
+from logging import getLogger
+from typing import List
+
+from fastapi import APIRouter, Body, Depends, Request, Security, status
+from fastapi.security import HTTPBasic
+from sqlalchemy.orm import Session
+
+import oauth.api.endpoints as endpoints
+from oauth.api.exceptions import (
+    AuthenticationException,
+    ClientNotFoundException,
+    InvalidScopeException,
+)
+from oauth.api.models import (
+    AccessToken,
+    ClientCreatedResponse,
+    OAuth2ClientCredentialsRequestForm,
+)
+from oauth.api.utils import verify_oauth_client
+from oauth.database.models import ClientDetail
+from oauth.scopes import (
+    CLIENT_CREATE,
+    CLIENT_DELETE,
+    CLIENT_READ,
+    CLIENT_UPDATE,
+    SCOPE_READ,
+    SCOPES,
+)
+
+logger = getLogger(__name__)
+router = APIRouter(prefix=endpoints.OAUTH_PREFIX, tags=["OAuth"])
+
+
+@router.post(endpoints.TOKEN, response_model=AccessToken)
+async def acquire_access_token(
+    request: Request,
+    form_data: OAuth2ClientCredentialsRequestForm = Depends(),
+    db: Session = Depends(),
+) -> AccessToken:
+    """
+    Given a set of credentials, returns an access token.
+    """
+
+    basic_credentials = await HTTPBasic(auto_error=False)(request)
+
+    if form_data.client_id and form_data.client_secret:
+        client_id = form_data.client_id
+        client_secret = form_data.client_secret
+    elif basic_credentials:
+        client_id = basic_credentials.username
+        client_secret = basic_credentials.password
+    else:
+        raise AuthenticationException()
+
+    client_detail = ClientDetail.get(db, client_id=client_id)
+
+    if client_detail is None:
+        raise AuthenticationException()
```

Comment:
This should probably be a `ClientNotFoundException`. I added that type after this code already existed for a while.

Reply:
That makes sense.

Reply:
@pattisdr in this instance, I think you're correct that a `ClientNotFoundException` would be less secure, as it would potentially foster a brute-force credential attack. However, the `AuthenticationException` neither specifies whether the credentials were missing or invalid, nor whether the client ID or the client secret were incorrect - it simply responds with a `401 Unauthorized` and the string `Missing or invalid credentials`. I believe this is a secure response, and so no changes here should be made. Thoughts?